examples/step-44/step-44.cc

```diff
@@ -2001,16 +1998,18 @@ namespace Step44
       {
         fe_values_ref.reinit(cell);
 
-        PointHistory<dim> *lqph =
-          reinterpret_cast<PointHistory<dim>*>(cell->user_pointer());
-
-        Assert(lqph >= &quadrature_point_history.front(), ExcInternalError());
-        Assert(lqph <= &quadrature_point_history.back(), ExcInternalError());
+        // In contrast to that which was previously called for,
+        // in this instance the quadrature point data is specifically
+        // non-modifiable since we will only be accessing data.
+        const std::vector<std::shared_ptr<const PointHistory<dim> > > lqph =
+            const_cast<const CellDataStorage<typename Triangulation<dim>::cell_iterator,
+                                             PointHistory<dim> >&>(quadrature_point_history).get_data(cell);
```

Comment:
Since @davydden asked me to join in, I will.

First, I think the `const_cast` is a bit awkward, but if properly explained in the comment ahead of it, I agree that it is a learning/teaching opportunity one can use. I wouldn't have a problem with it.

But, second: This happens in a function called `get_error_dilation()`, if I see this right. Functions that are called `get_*` should always be `const` because, well, they *get* something, not modify something. That would in fact solve the problem here.

In the comments above talk about possibly making the local assembler function `const` as well. I think that's generally a good strategy. Conceptually, the local assemblers used in `WorkStream` need to be able to run in parallel without using mutices. In fact, their whole design is such that the local assemblers do local work *and write their results into a `CopyData` object*, rather than modifying any global state of the object of which they are a member. This means, that the best way to ensure that they can run in parallel is to actually make them `const`. It is the duty of the `copy_local_to_global()` function to copy this data into the global object, which it can do because it is guaranteed to run only once in parallel. So making the local assembler functions and everything it calls `const` is generally a good idea as well.